### PR TITLE
feat: LT-945 - Attach documents to cases

### PR DESCRIPTION
**Pull request by jfaracik**
* Added documents and case document models
* Added serialisers for said models
* Added endpoints to add document metadata
* Added views for displaying all documents on a case, ordered by most recent

**TODO**
* Add API tests
* Add ClamAV

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -16,6 +16,9 @@ python-json-logger = "==0.1.11"
 parameterized = "*"
 drf-yasg = "*"
 prospector = "==1.1.6.2"
+boto3 = "*"
+django-pglocks = "*"
```
Comment (markm5A8Z9F43913A) [inline Pipfile]: You might want to specify the versions here.

File: cases/serializers.py [hunk 2]
```diff
@@ -1,12 +1,14 @@
 from rest_framework import serializers
-from rest_framework.relations import PrimaryKeyRelatedField
 
 from applications.serializers import ApplicationBaseSerializer
-from cases.models import Case, CaseNote, CaseAssignment
+from cases.models import Case, CaseNote, CaseAssignment, CaseDocument
+from conf.settings import ASYNC_DOC_PREPARE
 from content_strings.strings import get_string
+# from documents.serializers import DocumentSerializer
```
Comment (markm5A8Z9F43913A) [inline cases/serializers.py]: Why is this line still here?

File: documents/av_scan.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
+import boto3
+from contextlib import closing
+from logging import getLogger
+
+import requests
+from django.conf import settings
+from django.utils.timezone import now
+from django_pglocks import advisory_lock
+from requests_toolbelt.multipart.encoder import MultipartEncoder
+# from raven.contrib.django.raven_compat.models import client
```
Comment (markm5A8Z9F43913A) [inline documents/av_scan.py]: Is this still needed?

File: documents/models.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        try:
+            # self.update_md5_checksum()
+            self.scan_for_viruses()
+            if self.safe is False:
+                self.delete()
+            return bool(self.checksum) and self.safe
+        except Exception:
+            return False
+
+    # def update_md5_checksum(self):
```
Comment (markm5A8Z9F43913A) [inline documents/models.py]: Please clean up any commented out/unused code unless there is a really good reason for it to be there.

File: cases/libraries/get_case.py [hunk 5]
```diff
@@ -10,4 +9,24 @@ def get_case(pk):
...
+
+
+def get_case_document(case: Case, s3_key: str):
+    """
+    Returns a case or returns a 404 on failure
+    """
+    try:
+        return CaseDocument.objects.get(case=case, s3_key=s3_key)
+    except CaseDocument.DoesNotExist:
+        raise NotFoundError({'document': 'Queue not found'})
```
Comment (cstobart) [inline cases/libraries/get_case.py]: Put error message in strings.json

File: documents/tests.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
+
+# Create your tests here.
```
Comment (markm5A8Z9F43913A) [inline documents/tests.py]: TODO

File: Procfile [hunk 7]
```diff
@@ -1 +1,2 @@
 web: python manage.py migrate && ./manage.py runserver 0.0.0.0:$PORT
+worker: python manage.py process_tasks
```
Comment (markm5A8Z9F43913A) [inline Procfile]: Add this to docker as well, update documentation

File: conf/settings.py [hunk 8]
```diff
@@ -17,12 +17,11 @@
     LOG_LEVEL=(str, 'INFO'),
 )
 
-
 # Quick-start development settings - unsuitable for production
 # See https://docs.djangoproject.com/en/2.1/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = '(%0hafx7+lsw4m6n(t)h!#sje$n$er9&z4hrfewm%&64=4mhy9' # noqa
+SECRET_KEY = '(%0hafx7+lsw4m6n(t)h!#sje$n$er9&z4hrfewm%&64=4mhy9'  # noqa
```
Comment (markm5A8Z9F43913A) [inline conf/settings.py]: SECURITY WARNING: keep the secret key used in production secret! Also this doesn't seem to be used anywhere.

File: local.env [hunk 9]
```diff
@@ -4,4 +4,20 @@ PORT=8100
...
+AWS_STORAGE_BUCKET_NAME=''
+AWS_REGION=''
+
+# AV
+AV_SERVICE_URL=''
+AV_SERVICE_USERNAME=''
+AV_SERVICE_PASSWORD=''
+
+PYTHONUNBUFFERED=1
+BACKGROUND_TASK_RUN_ASYNC=True
```
Comment (markm5A8Z9F43913A) [inline local.env]: Newline

File: docker.env [hunk 10]
```diff
@@ -5,3 +5,18 @@ ALLOWED_HOSTS='["*"]'
...
+AWS_SECRET_ACCESS_KEY=''
+AWS_STORAGE_BUCKET_NAME=''
+AWS_REGION=''
+
+# AV
+AV_SERVICE_URL=''
+AV_SERVICE_USERNAME=''
+AV_SERVICE_PASSWORD=''
+
+BACKGROUND_TASK_RUN_ASYNC=False
```
Comment (markm5A8Z9F43913A) [inline docker.env]: True

File: cases/tests/tests_case_documents.py [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        self.create_case_document(case=self.case, user=self.user, name='doc1.pdf')
+        self.create_case_document(case=self.case, user=self.user, name='doc2.pdf')
+
+        response = self.client.get(self.url, **self.gov_headers)
+        response_data = response.json()
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(len(response_data['documents']), 2)
+
+    #still work in progress
```
Comment (cstobart) [inline cases/tests/tests_case_documents.py]: wrap these in a TODO if you plan to come back to them

File: cases/urls.py [hunk 12]
```diff
@@ -9,5 +9,9 @@
     path('<uuid:pk>/case_notes/', views.CaseNoteList.as_view(), name='case_notes'),
     # ex: /cases/<uuid:pk>/activity/
     # ex: /cases/<uuid:pk>/activity/?fields=activity,status
-    path('<uuid:pk>/activity/', views.ActivityList.as_view(), name='activity'),
+    path('<uuid:pk>/activity/', views.CaseActivity.as_view(), name='activity'),
```
Comment (cstobart) [inline cases/urls.py]: Nice

File: conf/settings.py [hunk 13]
```diff
@@ -15,14 +15,14 @@
...
+    BACKGROUND_TASK_ENABLED=(bool, False),
 )
 
-
 # Quick-start development settings - unsuitable for production
 # See https://docs.djangoproject.com/en/2.1/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = '(%0hafx7+lsw4m6n(t)h!#sje$n$er9&z4hrfewm%&64=4mhy9' # noqa
+SECRET_KEY = '(%0hafx7+lsw4m6n(t)h!#sje$n$er9&z4hrfewm%&64=4mhy9'  # noqa
```
Comment (cstobart) [inline conf/settings.py]: errrrm

**Pull request merged**